src/operation/insert/mod.rs

```diff
@@ -3,62 +3,103 @@ mod test;
 
 use std::collections::HashMap;
 
+use bson::{oid::ObjectId, Bson};
+use serde::Serialize;
+
 use crate::{
     bson::{doc, Document},
     bson_util,
     cmap::{Command, CommandResponse, StreamDescription},
-    error::{ErrorKind, Result},
+    error::{BulkWriteFailure, Error, ErrorKind, Result},
     operation::{append_options, Operation, Retryability, WriteResponseBody},
     options::{InsertManyOptions, WriteConcern},
     results::InsertManyResult,
     Namespace,
 };
 
 #[derive(Debug)]
-pub(crate) struct Insert {
+pub(crate) struct Insert<T> {
     ns: Namespace,
-    documents: Vec<Document>,
+    documents: Vec<T>,
+    inserted_ids: Vec<Bson>,
     options: Option<InsertManyOptions>,
 }
 
-impl Insert {
+impl<T> Insert<T> {
     pub(crate) fn new(
         ns: Namespace,
-        documents: Vec<Document>,
+        documents: Vec<T>,
         options: Option<InsertManyOptions>,
     ) -> Self {
         Self {
             ns,
             options,
-            documents: documents
-                .into_iter()
-                .map(|mut d| {
-                    bson_util::add_id(&mut d);
-                    d
-                })
-                .collect(),
+            documents,
+            inserted_ids: vec![],
         }
     }
+
+    fn is_ordered(&self) -> bool {
+        self.options
+            .as_ref()
+            .and_then(|options| options.ordered)
+            .unwrap_or(true)
+    }
 }
 
-impl Operation for Insert {
+impl<T: Serialize> Operation for Insert<T> {
     type O = InsertManyResult;
     const NAME: &'static str = "insert";
 
-    fn build(&self, _description: &StreamDescription) -> Result<Command> {
+    fn build(&mut self, description: &StreamDescription) -> Result<Command> {
+        let mut docs: Vec<Document> = vec![];
+        let mut size = 0;
+
+        for (i, d) in self
+            .documents
+            .iter()
+            .take(description.max_write_batch_size as usize)
+            .enumerate()
+        {
+            let mut doc = bson::to_document(d)?;
+            let id = doc
+                .entry("_id".to_string())
+                .or_insert_with(|| {
+                    self.inserted_ids
+                        .get(i)
+                        .cloned()
+                        .unwrap_or_else(|| Bson::ObjectId(ObjectId::new()))
+                })
+                .clone();
+
+            let doc_size = bson_util::array_entry_size_bytes(i, &doc);
+
+            if (size + doc_size) <= description.max_bson_object_size as u64 {
+                if self.inserted_ids.len() <= i {
```

Comment:
only populated the `inserted_ids` vector if this is the first time `build` is being called.